File: nlox/Lox/LoxFunction.cs
class LoxFunction : LoxCallable
{
  private Function declaration;
  private Environment closure;
  private bool isInitializer;

  public LoxFunction(Function declaration, Environment closure, bool isInitializer)
  {
    this.declaration = declaration;
    this.closure = closure;
    this.isInitializer = isInitializer;
  }

  public LoxFunction Bind(LoxInstance instance)
  {
    var environment = new Environment(closure);
    environment.Define("this", instance);
    return new LoxFunction(declaration, environment, isInitializer);
  }

  public int Arity => declaration.Params.Count;

  public object? Call(Interpreter interpreter, List<object?> arguments)
  {
    var environment = new Environment(closure);
    for (int i = 0; i < declaration.Params.Count; i++)
    {
      environment.Define(declaration.Params[i].Lexeme, arguments[i]);
    }

    try
    {
      interpreter.Execute(declaration.Body, environment);
    }
    catch (ReturnException retVal)
    {
      if (isInitializer)
      {
        return closure.GetAt(0, "this");
      }

      return retVal.Value;
    }

    if (isInitializer)
    {
      return closure.GetAt(0, "this");
    }
    
    return null;
  }

  public override string ToString() => $"<fn {declaration.Name.Lexeme} >";
}

File: tslox/src/loxfunction.ts
import { Function } from './stmt';
import Environment from './environment';
import { LoxCallable } from './loxcallable';
import Interpreter from './interpreter';
import ReturnException from './return';
import { LoxInstance } from './loxinstance';

export default class LoxFunction implements LoxCallable {
  constructor(
    private readonly declaration: Function,
    private readonly closure: Environment,
    private readonly isInitializer: boolean
  ) {
    this.arity = this.declaration.params.length;
  }

  bind(instance: LoxInstance) {
    const env = new Environment(this.closure);
    env.define("this", instance);
    return new LoxFunction(this.declaration, env, this.isInitializer);
  }

  readonly arity: number;

  call(interpreter: Interpreter, args: unknown[]): unknown {
    const env = new Environment(this.closure);
    for (let i = 0; i < this.declaration.params.length; i++) {
      env.define(this.declaration.params[i].lexeme, args[i]);
    }

    try {
      interpreter.executeBlock(this.declaration.body, env);
    } catch (e) {
      if (e instanceof ReturnException) {
        if (this.isInitializer) {
          return this.closure.getAt(0, "this");
        }

        return e.value;
      }
      throw e;
    }

    if (this.isInitializer) {
      return this.closure.getAt(0, "this");
    }
    
    return null;
  }

  toString = () => `<fn ${this.declaration.name.lexeme}>`;
}
